File: src/main/java/ru/ioleksiv/telegram/bot/api/annotations/filter/secondary/ShippingAddressFilter.java
package ru.ioleksiv.telegram.bot.api.annotations.filter.secondary;

import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.StringFilter;
import ru.ioleksiv.telegram.bot.api.model.annotation.AnnotationState;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Documented
public @interface ShippingAddressFilter {

    AnnotationState value() default AnnotationState.ON;

    StringFilter countryCode() default @StringFilter(AnnotationState.OFF);

    StringFilter state() default @StringFilter(AnnotationState.OFF);

    StringFilter city() default @StringFilter(AnnotationState.OFF);

    StringFilter streetLine1() default @StringFilter(AnnotationState.OFF);

    StringFilter streetLine2() default @StringFilter(AnnotationState.OFF);

    StringFilter postCode() default @StringFilter(AnnotationState.OFF);

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/annotations/filter/composite/ShippingAddressFilter.kt
package io.github.oleksivio.tl.kbot.core.annotations.filter.composite

import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.StringFilter
import io.github.oleksivio.tl.kbot.core.model.annotation.AnnotationState

@Retention
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@MustBeDocumented
annotation class ShippingAddressFilter(
    val status: AnnotationState = AnnotationState.ON,
    val validator: Array<String> = [],
    val countryCode: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val state: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val city: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val streetLine1: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val streetLine2: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val postCode: StringFilter = StringFilter(
        status = AnnotationState.OFF
    )
)
